[CBRD-25888] Performance Enhancement for XASL Cache Memory Management

sohee-dgist opened the pull request:
<http://jira.cubrid.org/browse/CBRD-25888>

### Purpose

XASL Cache의 메모리 사용량을 관리하는 로직과 Plan이 클 경우 Clone을 생성하지 않는 로직을 추가하였다.
Xasl Cache의 Plandump에 메모리 사용량을 보이도록 하여 User 입장에서 쉽게 Memory 사용량을 추적할 수 있도록 한다.

### Implementation

1. Hard Limit 추가
2. Soft Limit이상이면 Clear하는 로직 추가
3. Memory 사용량 관련 추적 변수 활성화
4. Plan max size보다 클 경우 clone을 바로 decache하는 로직 추가

### Remarks

hashmap의 iteration시 iter를 초기화 할때 bucket index의 초기화가 이루어지지 않는 부분을 별도로 수정하였다.
xasl cache는 session view와 session variable과 연관이 있으므로,
xasl cache를 아주 작게 설정하거나, invalidate할 경우 동작하지 않는다.

session variable은 result cache만 꺼도 동작하지 않는다.

다음의 내용이 메뉴얼에 추가될 예정이다.

Commit:
(feature) xasl cache memory 관리 로직 & memory dump 추가

- xasl cache memory 관리 로직 추가

  int soft_limit;
   -> hard limit의 0.8되는 size로 설정
  int hard_limit;
  int max_plan_size;
  volatile INT32 memory_usage_cache;
   -> cache memory 사용량 중 plan을 저장하는 부분
  volatile INT32 memory_usage_clone;
   -> cache memory 사용량 중 plan을 복사하는 부분

- memory가 다 차도 cache out 되도록 함.
- hard limit - soft limit 보다 큰 plan은 저장되지 못하도록 함.

- dump 에 추가

  entry의 메모리 사용량이 보이도록 함. (다만 아직 clone을 얼마만큼 사용하는지 확인 불가)
  전체 메모리 사용량이 보이도록 함.

Commit:
(오류해결) 클론사이즈 계산 로직 수정

decache 하위에서 clonesize를 계산하여 줄이도록 함.

Commit:
(refactor) pt_ 함수 parser support로 이동

Commit:
(feature) xasl cache clone size 구하는 로직 변경

- xasl cache clone size중 alloc buf의 사이즈는 사용하면 안 됨
- alloc_buf는 남은 사이즈를 의미함.

(feature) xasl cache memory cleanout 로직 변경

- candidate 지우는 로직 변경.
- count로 지우는 로직은 더이상 사용하지 않음.

Commit:
(refactor) bug fix용 임시 커밋

- xasl cache cleanout 매커니즘 테스트시 entry transction id관련 버그

Commit:
(feature) lock free iterator관련 오류 해결

Commit:
(bugfix) xasl cache clone size 구하는 로직 수정
추후 retire 할 때 clone이 제거되므로, 해당 시점에 원래는 soft limit을 낮추는
것이 맞으나, 해당 방법으로는 계속해서 clone이 쌓여 soft limit을 넘어 memory
cleanout이 계속하여 일어난다.
cleanout시에 clone역시 지우도록 변경되었다.

Commit:
Revert "(refactor) pt_ 함수 parser support로 이동"

This reverts commit 023140cda1791d0af26c8d4fb9c22cd91dbfea09.

# 제목: 변경 요약 (50자 이내)  (refactor, hotfix, feature)

# 본문: 변경 사항에 대한 자세한 설명
# - 주요 변경 사항 1
# - 주요 변경 사항 2

# 참고: 이 라인 아래의 내용은 커밋 메시지에서 제거됩니다.

Commit:
Merge remote-tracking branch 'upstream/develop' into CBRD_25888

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

Commit:
(bugfix) temp 변수 설정 잘못된 것 정상적으로 설정

- soft limit 설정 누락됨
- hard limit temp 변수 원상 복구

Commit:
(test) test용 임시 커밋

temp size 수정정

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

Commit:
(bugfix) xasl cache parameter 원상 복구

xasl cache는 session view와 session variable과 연관이 있으므로,
xasl cache를 아주 작게 설정하거나, invalidate할 경우 동작하지 않는다.

session variable은 result cache만 꺼도 동작하지 않는다.

다음의 내용을 메뉴얼에 추가할 것.

shparkcubrid commented:
테스트 케이스 등록해주세요. 소스코드 리뷰시 테스트 케이스도 리뷰 대상입니다.

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -514,13 +538,12 @@ xcache_entry_uninit (void *entry)
       XASL_ID_SET_NULL (&xcache_entry->xasl_id);
 
       /* Free XASL clones. */
-      assert (xcache_entry->n_cache_clones == 0
```
삭제한 이유가 있을까요?

sohee-dgist replied:
해당 경우는

xasl cache가 아주 클때 clone을 사용하는 옵션이 켜져 있더라도, clone을 바로 끝나자마자 decache하는 로직이 존재하므로, 해당 루트를 타게 됩니다.

assert가 걸린 이유를 사실 잘 모르겠지만, 일단 clone을 사용하도록 켜져 있는데 clone을 사용하지 않는 경우 assert를 낸 것 같습니다.

sohee-dgist replied:
그래서 신규 기능이 추가되었으므로, 가능한 시나리오로 보고 추가 기능에 맞추어 assert 역시 제거하였습니다.

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -1042,6 +1065,10 @@ xcache_find_xasl_id_for_execute (THREAD_ENTRY * thread_p, const XASL_ID * xid, X
     stx_map_stream_to_xasl (thread_p, &xclone->xasl, use_xasl_clone, (*xcache_entry)->stream.buffer,
 			    (*xcache_entry)->stream.buffer_size, &xclone->xasl_buf);
+  INT32 xasl_clone_size =
+    sizeof (xasl_unpack_info) + sizeof (XASL_NODE) + UNPACK_SCALE * (*xcache_entry)->stream.buffer_size;
```
UNPACK_SCALE * stream.buffer_size <== 이렇게 계산하는 이유가 무엇일까요?

sohee-dgist replied:
실제 캐시 클론 자체가 db private alloc으로 해당 부분만큼 alloc 받고 free하지 않은 상태로 사용합니다.

shparkcubrid replied:
최초 크기를 그렇게 설정하고 stx_alloc_struct()을 보면 넘었을 경우 추가로 할당하고 있습니다. 실제로 넘는 경우가 있을까 싶기는 한데요. xasl_unpack_info에 total_size와 같이 추가해서 정확한 크기를 측정하는 방안도 있겠지만 크게 의미는 없어보입니다. 그렇다면 관련해서 예상값이라고 적절하게 주석을 추가하는 것이 좋아 보입니다.

shparkcubrid replied:
그리고 함수로 만들어서 다른 루틴과 같이 사용하면 가독성이 좋아질 것으로 생각됩니다.

sohee-dgist replied:
해당 코드 확인 하였습니다. 그런데 unpack scale이 넘는 경우가 없기도 하고 추적할 방법이 없으므로, 주석을 다는 것이 좋을 것 같습니다.

sohee-dgist replied:
xcache_entry_get_clonesize 라고 해서 entry를 인자로 받아 clone의 전체 사이즈를 계산하는 함수는 존재하는데, 아닌 경우의 함수도 한벌 만들어 반영하도록 하겠습니다.

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -2320,6 +2457,13 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
 	  perfmon_inc_stat (thread_p, PSTAT_PC_NUM_DELETE);
 	  ATOMIC_INC_32 (&xcache_Entry_count, -1);
+	  ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (candidate.xcache));
+	  while (candidate.xcache->n_cache_clones > 0)
```
이전에는 clone 삭제가 누락된 것인가요?

sohee-dgist replied:
이전에는 clone 삭제를 나중에 하였었습니다. 주석에 보면  여기에서 해도 괜찮다고 하여서 이곳에서 하는 것으로 변경하였습니다.

shparkcubrid replied:
실행중인 clone이 있을 것을 대비해 xcache_unfix()시에 삭제하는 것으로 보이는데요. 그렇다면 삭제하는 루틴은 제거되어야 겠습니다. 
아래 분석내용은 소스코드를 보고 예측한 것이고 디버깅한 것은 아닙니다. 정확하지 않을 수 있습니다.
```
xcache_cleanup ()
cache_flag = XCACHE_ENTRY_CLEANUP

xcache_Hashmap.erase
xcache_compare_key ()
cache_flag = XCACHE_ENTRY_DELETED_BY_ME

xcache_unfix()
if (XCACHE_ENTRY_DELETED_BY_ME)
  xcache_clone_decache ()
```

sohee-dgist replied:
아닙니다.  예상 시나리오 테스트와 설명을 적는데 좀 오래 걸렸네요.

xcache clone의 decache의 경우, 예전에 이미 안전한 것을 확인했습니다.

왜냐하면 이미 clone이 사용되는 시점에는 (xasl_cache 1039~)

	  (void) pthread_mutex_lock (&(*xcache_entry)->cache_clones_mutex);
	  assert ((*xcache_entry)->n_cache_clones <= xcache_Max_clones);
	  if ((*xcache_entry)->n_cache_clones > 0)
	    {
	      /* A clone is available. */
	      *xclone = (*xcache_entry)->cache_clones[--(*xcache_entry)->n_cache_clones];
	      (void) pthread_mutex_unlock (&(*xcache_entry)->cache_clones_mutex);
	      
에서 이미 xcache entry에서 cache clone을 옮겨버린 상태이기 때문입니다.
추후 xasl clone retire에서 제거되는 로직과는 별도로 잘 돌아갑니다.


1) 다른 시나리오의 경우 

예) 그렇다면 xasl clone retire 시에 이미 cleanup이 진행되서 entry가 invalidated되어 있으면 어떡하는가?
xasl cache entry가 삭제된 것도 아니고 이미 잘 freelist에 들어가 있으므로 괜찮다. ( 제 로직 변경과 별개이므로 더 자세한 것은 잘.. 일단 devel에서 잘 돌아가면 제 것에서도 잘 돌아가게 됩니다.  이 부분 미 변경되었으므로... )

2) 다른 검증 절차를 위해 테스트한 시나리오

two thread scenario

(1) background process에서 plandump를 전부 지울 때 혹은 지울 때 xcache에 접근하도록 시도하면.
pthread mutex로 잡아 놓으므로, 애초에 다른 것이 끝나기 전 까지 실행되지 않음.
반대의 경우도 마찬가지.

shparkcubrid replied:
관련해서 조금더 분석해 보았는데 cache_clones_mutex 없이 삭제하는 것은 조금 위험해 보입니다.
기존의 xcache_clone_decache을 호출하는 함수는 cache_clones_mutex를 사용하지 않지만 cache_flag가 0인 것을 보장합니다. cache_flag가 0이라는 것은 clone이 사용되지 않음을 의미합니다.

- cache_clones_mutex 사용 
xcache_find_xasl_id_for_execute  <== 질의 실행시 clone 사용
xcache_retire_clone                <== 질의 실행후 clone 반환

- xcache_clone_decache을 호출하는 함수
xcache_entry_uninit
xcache_unfix
xcache_invalidate_entries

- cache_flag 변경
xcache_compare_key <== cache_flag + 1
xcache_unfix <== cache_flag - 1


```
while (candidate.xcache->n_cache_clones > 0)
{
      xcache_clone_decache (thread_p, &candidate.xcache->cache_clones[--candidate.xcache->n_cache_clones],
   		    	candidate.xcache);
}
```
위 루틴과 xcache_retire_clone()이 동시에 수행되면 문제가 될 수 있어 보입니다.

sohee-dgist replied:
다시 검토해 보았는데요, cache clones mutex 는 필요한게 맞습니다. 그 방향으로 수정 진행하도록 하겠습니다.

이곳에서 지워야 하는 이유는 메모리의 소프트 리밋을 넘기고 나서 cleanout하는 것인데 (oom 방지)
실제 메모리를 반환하지 않는 것이 logic상 이상하다고 보기 때문이었습니다.

mutex사용 하는것으로 바꾸어 반영하겠습니다.

sohee-dgist replied:
반영 완료 되었습니다.

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -1378,13 +1406,11 @@ xcache_insert (THREAD_ENTRY * thread_p, const compile_context * context, XASL_ST
       return NO_ERROR;
     }
 
-  xcache_check_logging ();
```
실수로 삭제한 것 같습니다.

sohee-dgist replied:
복원 하였습니다

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -1643,6 +1670,7 @@ xcache_insert (THREAD_ENTRY * thread_p, const compile_context * context, XASL_ST
       if (xcache_need_cleanup () != XCACHE_CLEANUP_NONE && xcache_Cleanup_flag == 0)
 	{
 	  /* Try to clean up some of the oldest entries. */
+	  /* TODO ! */
```
의미 없는 주석입니다.

sohee-dgist replied:
제거 완료하였습니다. 임시로 사용한 뒤 안 남겨 두었던 것 같군요....

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -55,6 +55,8 @@
 
 #define XCACHE_ENTRY_FIX_COUNT_MASK	    ((INT32) 0x00FFFFFF)
 
+#define UNPACK_SCALE 3		// change stx_init_xasl_unpack_info too
```
stx_init_xasl_unpack_info () 와 동일한 define을 사용하는 것이 좋을 것 같습니다. 예를 들면 xasl.h에 추가하는 것입니다.

sohee-dgist replied:
그게 맞을 것 같네요 반영 하겠습니다.

Commit:
(bugfix, review) 코드리뷰 제안사항 반영

- xcache_entry_get_one_clonesize 함수 추가
- xcache drop시 정상적으로 memory 사용량 반영되지 않는 문제 개선
- 기타 코드 리뷰 제안사항 반영 (누락 된 파일 복구 등)

Commit:
(review) 코드리뷰 제안사항 2차 반영

- xasl unpack scale 다른 헤더로 옮기기
- xasl_cache.h 에서 주석 추가

sohee-dgist commented:
> 테스트 케이스 등록해주세요. 소스코드 리뷰시 테스트 케이스도 리뷰 대상입니다.

cte test 중 testcase를 추가하기 위해서, 

sql에다가 임시로 shell 명령어 쓸려고 ;shell 후 명령어 입력 후 quit을 쓰는데 정상적으로 작동 하지 않습니다. 무슨 오류인지는 잘 모르겠습니다. 오류는 제가 확인해보고 있으니 이 부분에 대해서는 도움이 필요 없는데, 혹시 내부에서 shell명령어를 쓰는 test에 대한 sql test이미 추가된 예제가 있는 것에 대해 알고 계시다면 확인 부탁드립니다.



또한, 테스트 시나리오의 경우 작은 환경에서 이루어져야 하는데 이게 부팅 시 파라미터다 보니 바꾸는 방법을 모르겠습니다.

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

sohee-dgist committed:
(recover) xasl cache limit temp 변수 복원

youngjinj commented:
> > 테스트 케이스 등록해주세요. 소스코드 리뷰시 테스트 케이스도 리뷰 대상입니다.
> 
> cte test 중 testcase를 추가하기 위해서,
> 
> sql에다가 임시로 shell 명령어 쓸려고 ;shell 후 명령어 입력 후 quit을 쓰는데 정상적으로 작동 하지 않습니다. 무슨 오류인지는 잘 모르겠습니다. 오류는 제가 확인해보고 있으니 이 부분에 대해서는 도움이 필요 없는데, 혹시 내부에서 shell명령어를 쓰는 test에 대한 sql test이미 추가된 예제가 있는 것에 대해 알고 계시다면 확인 부탁드립니다.
> 
> 또한, 테스트 시나리오의 경우 작은 환경에서 이루어져야 하는데 이게 부팅 시 파라미터다 보니 바꾸는 방법을 모르겠습니다.

csql에서 실행할 때는 csql 세션 명령어를 사용해서 shell 명령어 실행이 가능합니다. sql 테스트의 경우에는 csql에서 실행되는 것이 아니라 jdbc를 통해 cub_cas에서 sql을 실행하기 때문에 csql 세션 명령어를 사용할 수 없습니다. 그래서 shell 명령어를 실행할 수 없습니다. shell 명령어 실행이 필요한 경우에는 shell 테스트 케이스로 추가하고 있습니다.

테스트 케이스를 만들 때 shell 스크립트로 만들어도 괜찮습니다. 가능하면 sql 테스트로 수행될 수 있도록 하는 것이 테스트 기회가 많지만, 어쩔 수 없는 경우에는 shell 테스트 케이스로 수행하는 것이 맞다고 생각합니다.

shell 테스트 케이스는 shell/_01_utility/_01_sqlx/bug_cubridsus2018/cases/bug_cubridsus2018.sh 테스트 케이스를 참고해주세요. 시스템 파라미터를 변경하면서 테스트를 수행하게 할 수 있습니다.

shparkcubrid commented:
> test 중 testcase를 추가하기 위해서,
> 
> sql에다가 임시로 shell 명령어 쓸려고 ;shell 후 명령어 입력 후 quit을 쓰는데 정상적으로 작동 하지 않습니다. 무슨 오류인지는 잘 모르겠습니다. 오류는 제가 확인해보고 있으니 이 부분에 대해서는 도움이 필요 없는데, 혹시 내부에서 shell명령어를 쓰는 test에 대한 sql test이미 추가된 예제가 있는 것에 대해 알고 계시다면 확인 부탁드립니다.

shell 테스트 케이스로 만들어도 되지만... 그 부분은 QA의 업무 영역으로 봐야 합니다. 여기서 테스트 케이스의 작성은 사람이 수행해서 확인할 수 있도록 작성하는 것입니다. 예를 들면 아래와 같습니다.
```
-- 파라메터 설정
cubrid.conf
max_plan_cache_entries = 10

--서버 시작
cubrid server start demodb

--plan 추가 확인
csql > select * from game where host_year = 2004;

]# cubrid plan dump demodb
---------------------------------------------------------
XASL Cache Memory Info:
  Memory Hard Limit:           1000.00 MB
  Current Memory (cache):      320.30 KB
  Current Memory (clone):      794.45 KB    <== 필요한 결과 부분만 작성
  Total Memory:                1114.75 KB
  Max Plan Size:               66.67 MB
  Usage Percent:               0.11%
...
---------------------------------------------------------
```

sohee-dgist commented:
> > test 중 testcase를 추가하기 위해서,
> > sql에다가 임시로 shell 명령어 쓸려고 ;shell 후 명령어 입력 후 quit을 쓰는데 정상적으로 작동 하지 않습니다. 무슨 오류인지는 잘 모르겠습니다. 오류는 제가 확인해보고 있으니 이 부분에 대해서는 도움이 필요 없는데, 혹시 내부에서 shell명령어를 쓰는 test에 대한 sql test이미 추가된 예제가 있는 것에 대해 알고 계시다면 확인 부탁드립니다.
> 
> shell 테스트 케이스로 만들어도 되지만... 그 부분은 QA의 업무 영역으로 봐야 합니다. 여기서 테스트 케이스의 작성은 사람이 수행해서 확인할 수 있도록 작성하는 것입니다. 예를 들면 아래와 같습니다.
> 
> ```
> -- 파라메터 설정
> cubrid.conf
> max_plan_cache_entries = 10
> 
> --서버 시작
> cubrid server start demodb
> 
> --plan 추가 확인
> csql > select * from game where host_year = 2004;
> 
> ]# cubrid plan dump demodb
> ---------------------------------------------------------
> XASL Cache Memory Info:
>   Memory Hard Limit:           1000.00 MB
>   Current Memory (cache):      320.30 KB
>   Current Memory (clone):      794.45 KB    <== 필요한 결과 부분만 작성
>   Total Memory:                1114.75 KB
>   Max Plan Size:               66.67 MB
>   Usage Percent:               0.11%
> ...
> ---------------------------------------------------------
> ```

테스트 케이스 추가되었습니다. CBRD-25888에 댓글로도 달려 있으나 이곳에도 추가합니다. 확인 바랍니다.

```
설정을 다음과 같이 하여 준다.

cubrid.conf
max_plan_cache_entries = 10
cubrid server start demodb

초기값 확인은 다음과 같이 한다.

– 초기값 확인
$ cubrid plandump demodb
--------------------------------------------------------
XASL Cache Memory Info:
Memory Hard Limit: 1.25 MB
--------------------------------------------------------

– 한 개의 쿼리 확인
csql >  select * from dual
$ cubrid plandump demodb
------------------------------------------------------------
XASL_ID = {
sha1 =

{ 82af18a4 44ac3e16 45a1dba2 258d2a99 46c17301 }
,
time_stored = 1741756722 sec, 285560 usec
}
Memory Usage: 1283.00 Bytes (0.1% of total cache)
Clone Memory Usage: 7544.00 Bytes (0.6% of total cache)
fix_count = 0
cache flags = 00000000
reference count = 2
time second last used = 1741756722
clone count = 1
sql info:
SQL_ID = 22f038079fc78
sql user text = select * from dual
sql hash text = select [dual].[dummy] from [dual] [dual]?193="ko_KR";194="ko_KR";249="Asia/Seoul";user=0|897|2
sql plan text =
Sequential scan(dual dual)
OID_LIST (count = 1):
OID = 0|194|13, LOCK = IS_LOCK, TCARD = 1
------------------------------------------------------------

– 초기화 확인

$ cubrid plandump -d demodb
------------------------------------------------------------
XASL Cache Memory Info:
Memory Hard Limit: 1.25 MB
Current Memory (cache): 0.00 KB
Current Memory (clone): 0.00 KB
Total Memory: 0.00 KB
Max Plan Size: 0.08 MB
Usage Percent: 0.00%
------------------------------------------------------------

– 아주 큰 사이즈의 쿼리 (상대적) 이 올때 클론 미생성 확인

csql > create table t111 (col1 int, col2 int, col3 int, col4 varchar(20), col5 numeric(20,10), col6 varchar(20), col7 varchar(20), col8 varchar(20));
csql > insert into t111
select
rownum,
rownum / 20,
1,
lpad (to_char (rownum / 1000), 20, '0'),
rownum / 100000,
lpad (to_char (rownum / 10000), 20, '0'),
lpad (to_char (rownum / 100), 20, '0'),
lpad (to_char (rownum / 10), 20, '0')
from db_class a, db_class b, db_class c, db_class d, db_class e
limit 5000;

------------------------------------------------------------
XASL_ID = {
sha1 =

{ ef0487b8 1fb8024c 0e0f28ce f18bb1e5 c010bb43 }
,
time_stored = 1741756914 sec, 222002 usec
}
Memory Usage: 81826.00 Bytes (6.2% of total cache)
Clone Memory Usage: 0.00 Bytes (0.0% of total cache)
fix_count = 0
cache flags = 00000000
reference count = 2
time second last used = 1741756914
clone count = 0

------------------------------------------------------------

– 길게 돌리던 중 다른 곳에서 드랍하기.

csql > insert into t111
select
rownum,
rownum / 20,
1,
lpad (to_char (rownum / 1000), 20, '0'),
rownum / 100000,
lpad (to_char (rownum / 10000), 20, '0'),
lpad (to_char (rownum / 100), 20, '0'),
lpad (to_char (rownum / 10), 20, '0')
from db_class a, db_class b, db_class c, db_class d, db_class e
limit 500000;

 
$ cubrid plandump -d demodb (남아 있는 것 확인할 것)

– 초기화
$ cubrid plandump -d demodb

– cleanup 시나리오 확인
sql test 중 해당 것을 입력
cubrid-testcases/medium/_02_xtests/cases/outerjoin1.sql at develop · CUBRID/cubrid-testcases
cubrid-testcases/medium/_02_xtests/cases/outerjoin2.sql at develop · CUBRID/cubrid-testcases

cubrid plandump demodb 

XASL Cache Memory Info:
Memory Hard Limit: 1.25 MB
Current Memory (cache): 187.80 KB
Current Memory (clone): 747.59 KB
Total Memory: 935.40 KB
Max Plan Size: 0.08 MB
Usage Percent: 73.08%

– cleanup 이 일어나기 직전
sql test중 해당 것을 입력
cubrid-testcases/sql/_06_manipulation/_01_select/_005_order_by/cases/1007.sql at develop · CUBRID/cubrid-testcases
밀리는 오류 날 수 있음 중요하지 않음.

– 올바르게 cache cleanup이 일어나는지 확인
$ cubrid plandump demodb

XASL cache
Stats:
Max size: 10
Current entry count: 45
Lookups: 345
Hits: 203
Miss: 142
Inserts: 105
Found at insert: 4
Recompiles: 4
Failed recompiles: 0
Deletes: 1
Fix: 207
Unfix: 312
Cache cleanups: 4
Deletes at cleanup: 55

XASL Cache Memory Info:
Memory Hard Limit: 1.25 MB
Current Memory (cache): 192.98 KB
Current Memory (clone): 637.85 KB
Total Memory: 830.83 KB
Max Plan Size: 0.08 MB
Usage Percent: 64.91%

$ cubrid plandump -d demodb

전부 0으로 초기화 되어있는지 확인.
```

xmilex-git commented on src/query/xasl_cache.c:
```diff
@@ -114,7 +114,11 @@ struct xcache
+  int hard_limit;
+  int max_plan_size;
+  volatile INT32 memory_usage_cache;
+  volatile INT32 memory_usage_clone;
```
```suggestion
  INT32 memory_usage_clone;
```
전역 memory_usage_cache, memory_usage_clone의 값을 변경할 때는 ATOMIC_INC_32 등의 함수를 사용합니다.
따라서 동시성이 보장된다고 볼 수 있다고 생각합니다.
만약 volatile 또한 동시성 보장을 위해 사용한 것이라면 빼는 것이 어떨까요?
다른 이유가 있어서 volatile을 사용했다면 그 이유를 알려주시면 감사하겠습니다.

sohee-dgist replied:
제거 완료하였습니다.

sohee-dgist committed:
(refactor) 무의미한 volatile 제거

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

shparkcubrid commented on src/query/xasl.h:
```diff
@@ -55,6 +55,8 @@
 struct binary_heap;
 #endif // SERVER_MODE || SA_MODE
 
+#define UNPACK_SCALE 3		// change stx_init_xasl_unpack_info too
```
```suggestion
#define UNPACK_SCALE 3		// Assumed memory ratio when unpacking stream to XASL
```

Commit:
(bugfix) unsaftey xasl cache cleanout logic 변경

sohee-dgist committed:
Update src/query/xasl.h

Co-authored-by: SE park <shpark@cubrid.com>

Commit:
(refactor) xasl cache indent 맞춰 수정

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -1998,17 +2086,46 @@ xcache_dump (THREAD_ENTRY * thread_p, FILE * fp)
   fprintf (fp, "Deletes at cleanup:	    %lld\n", (long long) XCACHE_STAT_GET (deletes_at_cleanup));
   /* add overflow, RT checks. */
 
+  /* Memory info */
```
위 2073라인의 Max_size는 삭제 혹은 적절하게 변경하는 것이 좋겠습니다.
```
fprintf (fp, "Max size: %d\n", xcache_Soft_capacity);
```

sohee-dgist replied:
넵 반영하겠습니다.

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -1803,10 +1828,12 @@ xcache_invalidate_entries (THREAD_ENTRY * thread_p, bool (*invalidate_check) (XA
 		    }
 		  delete_xids[n_delete_xids++] = xcache_entry->xasl_id;
 		}
+	      ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (xcache_entry));
```
xcache_Hashmap.erase시에 xcache_entry는 retired_list에 옮겨졌다가 freelist로 transport 할때 xcache_entry_uninit()에서 할당 해제 됩니다. 관련 루틴을 해당 함수로 옮기는 것이 더 정확해 보입니다.
```
xcache_Hashmap.erase
lf_freelist_retire
  lf_freelist_claim
    lf_freelist_transport
      xcache_entry_uninit
```
참고로 아래 코드를 수정하면 테스트시에 lf_freelist_transport()를 더 자주 호출합니다.
```
--- a/src/base/lock_free.c
+++ b/src/base/lock_free.c
@@ -225,7 +225,7 @@ lf_tran_system_init (LF_TRAN_SYSTEM * sys, int max_threads)
   sys->used_entry_count = 0;
   sys->global_transaction_id = 0;
   sys->min_active_transaction_id = 0;
-  sys->mati_refresh_interval = 100;
+  sys->mati_refresh_interval = 10;

   return NO_ERROR;
 }
```

sohee-dgist replied:
이렇게 바꿀 경우 memory 사용량이 바로 반영되지 않아서 계속해서 cleanout할려고 노력하는 현상이 나타납니다.
이미 cleanout을 예상하고 메모리 사용량을 옮겨놓는것이 낫지 않을까요?

xmilex-git commented on src/query/xasl_cache.c:
```diff
@@ -293,14 +310,20 @@ xcache_initialize (THREAD_ENTRY * thread_p)
   xcache_Soft_capacity = prm_get_integer_value (PRM_ID_XASL_CACHE_MAX_ENTRIES);
   xcache_Time_threshold = prm_get_integer_value (PRM_ID_XASL_CACHE_TIME_THRESHOLD_IN_MINUTES) * 60;
 
+  xcache_Hard_limit = xcache_Soft_capacity * 1024 * 128;
+  xcache_Soft_limit = xcache_Hard_limit * 0.8;
```
기존에 사용하던 max_plan_cache_entries 는 캐시 엔트리의 갯수를 제한했었는데요, 이 PR이 적용된 후에는 그 값을 설정해도 (해당 값) * 128KB 로 실제 메모리 크기 기반으로 제한합니다.
이제 plandump 등의 유틸리티를 사용하면 사용중인 메모리 크기를 안내하게 될텐데요.
이에 따라 max_plan_cache_size 와 같은 파라미터를 추가하여 크기를 기반으로 최대치를 설정할 수 있게끔 하면 어떨까요?

xmilex-git replied:
같은 기능의 파라미터로 두개의 파라미터가 존재할 시 혼란을 초래할 수 있고, 아래 버전과의 호환성을 생각할 때 새로운 파라미터를 추가하지 않는 것으로 회의때 논의하였습니다.

xmilex-git commented on src/query/xasl_cache.c:
```diff
@@ -293,14 +310,20 @@ xcache_initialize (THREAD_ENTRY * thread_p)
       return NO_ERROR;
     }
 
   xcache_Max_clones = prm_get_integer_value (PRM_ID_XASL_CACHE_MAX_CLONES);
```
max_plan_cache_clones 파라미터를 설정하면 그 값에 대해 max_plan_cache_entries 처럼 실제 메모리 크기로 치환(예 : 값 * 128KB)하여 관리되나요?

sohee-dgist replied:
max plan cache clone은 변경 예정입니다. 그냥 클론의 최대 수로...

클론 메모리+플랜 메모리를 다 포함하여 max plan cache entries로만  관리됩니다.

xmilex-git commented on src/query/xasl_cache.c:
```diff
@@ -1173,6 +1198,7 @@ xcache_unfix (THREAD_ENTRY * thread_p, XASL_CACHE_ENTRY * xcache_entry)
     }
 }
 
+
 /*
```
```suggestion

/*
```
해당 PR과 관련없는 newline 인것 같아서 남깁니다

sohee-dgist replied:
적용하겠습니다.

xmilex-git reviewed (commented):
newline을 임의로 추가하거나 제거하면 git history 추적 시 혼동을 줄 우려가 있습니다. 확인하고 원래 코드대로 바꿔주시면 좋을것 같습니다.

xmilex-git commented on src/query/xasl_cache.c:
```diff
@@ -108,13 +108,17 @@ struct xcache
+  int soft_limit;
+  int hard_limit;
+  int max_plan_size;
+  INT32 memory_usage_cache;
+  INT32 memory_usage_clone;
```
INT32의 최대값은 2,147,483,647입니다. xasl_cache로 2GB이상을 사용할 일이 있다면 오류 발생이 우려됩니다.

xmilex-git replied:
현재 max_plan_cache_entries는 상한값이 없습니다. 기본값은 1000이므로 128MB 정도이지만 16384 이상으로 더 늘린다면 문제가 될 수도 있겠네요.

xmilex-git replied:
```c
/* src/base/system_parameter.c, 3671 line */
  {PRM_ID_XASL_CACHE_MAX_ENTRIES,
   PRM_NAME_XASL_CACHE_MAX_ENTRIES,
   (PRM_FOR_CLIENT | PRM_FOR_SERVER | PRM_FORCE_SERVER),
   PRM_INTEGER,
   &prm_xasl_cache_max_entries_flag,
   (void *) &prm_xasl_cache_max_entries_default,
   (void *) &PRM_XASL_CACHE_MAX_ENTRIES,
   /* TODO: define a maximum value. We cannot have any size hash table. */
   (void *) NULL, (void *) NULL,
   (char *) NULL,
   (DUP_PRM_FUNC) NULL,
   (DUP_PRM_FUNC) NULL},
```
현재 상한선이 없는 파라미터라 적절한 조치가 필요할 듯 합니다.

sohee-dgist replied:
파라미터에 대해서 상한과 하한을 정하고 낮을 경우나 높을경우 bad parameters 적용하도록 하겠습니다.

xmilex-git replied:
100(500?)~16000 정도의 범위로 설정하는 것으로 회의 때 논의하였습니다.

sohee-dgist replied:
상한선 설정 완료 하였습니다. 하한은 적용하지 않았습니다.

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -1803,10 +1828,12 @@ xcache_invalidate_entries (THREAD_ENTRY * thread_p, bool (*invalidate_check) (XA
 		    }
 		  delete_xids[n_delete_xids++] = xcache_entry->xasl_id;
 		}
+	      ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (xcache_entry));
```
```suggestion
	      /* The entry is actually deleted in xcache_entry_uninit(), and may not be deleted immediately. 
	       * The reason for decrement here is to avoid unnecessary cleanup.
	       */
	      ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (xcache_entry));
```

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -2320,6 +2465,14 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
 	  XCACHE_STAT_INC (deletes_at_cleanup);
 	  perfmon_inc_stat (thread_p, PSTAT_PC_NUM_DELETE);
 	  ATOMIC_INC_32 (&xcache_Entry_count, -1);
+	  ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (candidate.xcache));
```
```suggestion
	  /* The entry is actually deleted in xcache_entry_uninit(), and may not be deleted immediately. 
	   * The reason for decrement here is to avoid unnecessary cleanup.
	   */
	  ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (candidate.xcache));
```

shparkcubrid replied:
그리고 위 2456라인의 주석은 삭제 혹은 적절하게 수정해 주세요.

shparkcubrid replied:
candidate가 꼭 삭제되지 않습니다. xcache_compare_key() 를 보면 cache_flag가 0인 경우에만 동일한 값으로 취급하고 삭제됩니다. fix된 entry는 삭제되지 않습니다. 오랜 기간 운영을 하다 보면 문제가 될 수 있겠네요.  xcache_entry_uninit()에 추가하는 것이 좋아 보입니다.

sohee-dgist commented on src/query/xasl_cache.c:
```diff
@@ -2320,6 +2465,14 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
 	  XCACHE_STAT_INC (deletes_at_cleanup);
 	  perfmon_inc_stat (thread_p, PSTAT_PC_NUM_DELETE);
 	  ATOMIC_INC_32 (&xcache_Entry_count, -1);
+	  ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (candidate.xcache));
```
넵 알겠습니다. uninit시 수행되지만 지금은 지워지는 것으로 보임 으로 수정하도록 하겠습니다!

shparkcubrid replied:
candidate가 꼭 삭제되지 않습니다. xcache_compare_key() 를 보면 cache_flag가 0인 경우에만 동일한 값으로 취급하고 삭제됩니다. fix된 entry는 삭제되지 않습니다. 오랜 기간 운영을 하다 보면 문제가 될 수 있겠네요. xcache_entry_uninit()에 추가하는 것이 좋아 보입니다.

shparkcubrid replied:
xcache_Hashmap.erase ()에서 삭제 대상이 없을 경우 false를 리턴하기 때문에 위 내용은 무시해주세요.

sohee-dgist committed:
xasl cache new line 임의 추가된 것 삭제

Co-authored-by: Ilhan <70995854+xmilex-git@users.noreply.github.com>

shparkcubrid commented on src/query/xasl_cache.c:
```diff
@@ -2320,6 +2465,14 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
+	      xcache_clone_decache (thread_p, &candidate.xcache->cache_clones[--candidate.xcache->n_cache_clones],
+				    candidate.xcache);
+	    }
+	  (void) pthread_mutex_unlock (&candidate.xcache->cache_clones_mutex);
```
이 이슈와는 별도로 잠재적인 문제가 있는 것 같아 공유합니다. 다른 리뷰어도 의견주세요.
2402 라인
```
if (candidate.xid.cache_flag > 0 || (candidate.xid.cache_flag & XCACHE_ENTRY_FLAGS_MASK))
  {
     continue;
  }
(void) bh_try_insert (bh, &candidate, NULL);
```
2448 라인
```
candidate.xid.cache_flag = XCACHE_ENTRY_CLEANUP;
```
cleanup 진행 중에 cache_flag가 다른 쓰레드로 인해서 변경될 수 있고, 그것이 무시될 수 있다는 생각입니다. 아래와 같이 flag 변경이 성공하면 candidate로 설정해야 한다는 생각입니다.

```
if (XCACHE_ATOMIC_CAS_CACHE_FLAG (entry_key, 0, XCACHE_ENTRY_CLEANUP))
	{
        }

shparkcubrid replied:
이 문제는 팀스에 기록하여 추후에 다른 이슈로 처리하도록 하겠습니다.

youngjinj commented on src/query/xasl_cache.c:
```diff
@@ -108,13 +108,17 @@ struct xcache
-
+  int soft_limit;
+  int hard_limit;
+  int max_plan_size;
```
이 값을 `xcache_entry_get_one_clonesize ()`와 비교하는데
저는 `plan`이라는 용어에서 혼란이 있습니다.
`max_clone_size`는 어떤가요?
혹시 `plan`이라는 용어를 사용해야 했던 이유가 있어서 공유해주시면
이 부분을 이해하는데 도움이 될 것 같습니다.

sohee-dgist replied:
잘못 비교한 걸로 보입니다 수정하겠습니다

youngjinj commented on src/query/xasl_cache.c:
```diff
@@ -1803,10 +1827,12 @@ xcache_invalidate_entries (THREAD_ENTRY * thread_p, bool (*invalidate_check) (XA
 		    }
 		  delete_xids[n_delete_xids++] = xcache_entry->xasl_id;
 		}
+	      ATOMIC_INC_32 (&xcache_Memory_usage_cache, -xcache_entry_get_entrysize (xcache_entry));
```
해당 XASL을 사용 중인 상태에서 `cubrid plandump -d demodb`를 반복 실행하면 `xcache_Memory_usage_cache`가 반복해서 감소하는 것으로 확인됩니다. `del_mark` 시 최초 1회만 감소하게 하거나 실제로 제거하는 부분에서 감소시켜야 하는 것으로 보입니다.
```sql
drop table if exists t1;
create table t1 (c1 int, c2 int);
insert into t1 values (1, 1);

-- max_plan_cache_entries=10
-- max_plan_cache_clones=10

/* 세션1 - 모니터링 */
watch -n 1 cubrid plandump demodb

/* 세션2 - 실행 후 세션3 진행 */
csql -u dba demodb -c 'select c2 from t1 where c1 = 1 and sleep (10) != 1'

/* 세션3 */
cubrid plandump -d demodb
cubrid plandump -d demodb
cubrid plandump -d demodb
```

sohee-dgist replied:
해결했습니다. 곧 기타 수정 후 푸시할 예정이니 푸시 후 한번 더 알림 드릴테니 확인 부탁드려요.

sohee-dgist replied:
해결했습니다.

Commit:
(Review) 코드리뷰 내용 반영

- new line 제거 및 복구
- plandump -d 시 일어날 수 있는 상황에 대한 오류 해결 완료

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

Commit:
(review) 코드리뷰 내용 반영 2

- 상한 추가

Commit:
(refactor) indent에 맞추어 수정

Commit:
(review) 코드리뷰 내용 반영 3

- 주석 수정

Commit:
(refactor) xasl cache plandump 수정

-- max size 삭제

Commit:
(review) 코드리뷰 반영사항 수정 4

- xasl cache del mark시 무조건 true인 상황에 확실히 반환하도록 수정

shparkcubrid commented on src/base/system_parameter.c:
```diff
@@ -1366,6 +1366,7 @@ static unsigned int prm_auto_restart_server_flag = 0;
 
 int PRM_XASL_CACHE_MAX_ENTRIES = 1000;
 static int prm_xasl_cache_max_entries_default = 1000;
+static int prm_xasl_cache_max_entries_upper = 10000;
```
쉘테스트 필요

sohee-dgist replied:
확인했습니다.

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

sohee-dgist commented:
/run all

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

sohee-dgist commented:
/run all

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

sohee-dgist commented:
/run all

Commit:
(commit) fix: xasl_cache.c 수정

- xasl_clone을 사용하지 않을 경우 메모리 사용량을 증가시키지 않도록 결정

sohee-dgist commented:
/run all

sohee-dgist committed:
Merge branch 'CUBRID:develop' into CBRD_25888

sohee-dgist commented:
/run all
